core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala

```diff
@@ -1068,7 +1063,17 @@ class TimeSeriesShard(val dataset: Dataset,
     try {
       val part: FiloPartition = getOrAddPartition(recordBase, recordOff, group, ingestOffset)
       if (part == OutOfMemPartition) { disableAddPartitions() }
-      else { part.asInstanceOf[TimeSeriesPartition].ingest(binRecordReader, overflowBlockFactory) }
+      else {
+        part.asInstanceOf[TimeSeriesPartition].ingest(binRecordReader, overflowBlockFactory)
+        activelyIngesting.synchronized {
```

Comment:
Instead of full synchronization, the use of a StampedLock is cheaper in the common case. This does require more changes, however.